Grading incompatible types

ChristopherChudzicki opened the pull request:
Provides configuration keys to customize what should happen when a student submits an answer to MatrixGrader that has the wrong shape.

ChristopherChudzicki commented:
@jolyonb This is ready for review. I will address any changes you think should happen to this, but other than that I probably won't do anything until July 28th. (In particular, let me know if you think different default settings would be better.)

jolyonb commented on mitxgraders/formulagrader/matrixgrader.py:
```diff
@@ -59,3 +80,31 @@ def check_response(self, answer, student_input, **kwargs):
+                expected_shape = get_description(comparer_params[0], detailed=detailed)
+                received_shape = get_description(student_input, detailed=detailed)
+                if not detailed and expected_shape == received_shape:
+                    msg = ("Exepected answer to be a {0}, but input is a {1}."
```
Expected

jolyonb commented on mitxgraders/formulagrader/matrixgrader.py:
```diff
@@ -59,3 +80,31 @@ def check_response(self, answer, student_input, **kwargs):
+                received_shape = get_description(student_input, detailed=detailed)
+                if not detailed and expected_shape == received_shape:
+                    msg = ("Exepected answer to be a {0}, but input is a {1}."
+                           "of incorrect shape.".format(expected_shape, received_shape))
```
Wait, did this just run on from the period directly?

ChristopherChudzicki replied:
Do you mean this?
```python
continued_string = ( "a"
                                   "b" )
continued_string == "ab" # True
```

But the period was a typing. Looking at this more closely, apparently I forgot `assert` in all of my tests! (except the error-raising ones). So that needs to be fixed.

jolyonb replied:
First line: "... but input is a {1}." (ends in a period).
Second line: "of incorrect shape."

I think you want to change the period to a space?

ChristopherChudzicki replied:
Yep, and it didn't break the tests because I forgot the asserts. =/

jolyonb commented on mitxgraders/formulagrader/matrixgrader.py:
```diff
@@ -59,3 +80,31 @@ def check_response(self, answer, student_input, **kwargs):
+                if not detailed and expected_shape == received_shape:
+                    msg = ("Exepected answer to be a {0}, but input is a {1}."
+                           "of incorrect shape.".format(expected_shape, received_shape))
+                msg = ("Exepected answer to be a {0}, but input is a {1}."
```
Expected again.

jolyonb commented:
Apart from some spelling/grammar, the implementation in `default_equality_comparer` is perfectly fine. However, can we take the try/except out of `default_equality_comparer` and wrap it around the call to the comparer, so that other comparers benefit from the same error handling?

ChristopherChudzicki commented:
Regarding
> However, can we take the try/except out of default_equality_comparer and wrap it around the call to the comparer, so that other comparers benefit from the same error handling?

I think you're right that we need a more generic solution here, but I'm not sure that this will work in general. In order to make appropriate error messages, we need to know what shape the answer should be. In the first version of this PR, the expected shape was determined from `comparer_params[0]` which, for the default comparer, is the actual answer. But for a general comparer, the shape of comparer_params[0] might not be the shape of the answer. (For example, for a comparer like `is_eigenvector_of`.)

So instead I moved the shape validation to a separate function, and this function is provided to the comparer function as part of `utils`.

## Also:
The original version of this PR had a number of issues that I've hopefully now resolved.

- types: fixed
- tests: In original PR I forgot to assert the comparisons, so several failing tests were not revealed. These have been fixed. Fixing the failing tests required:
    - `comparer_function`s can now return either `True`, `False`, `'partial'`, or a dictionary. This lets comparer messages pass specific failure messages.

jolyonb commented on mitxgraders/formulagrader/matrixgrader.py:
```diff
@@ -59,3 +85,78 @@ def check_response(self, answer, student_input, **kwargs):
+            utils.validate_shape(student_input, shape)
+        except ShapeError as err:
+            if self.config['answer_shape_mismatch']['is_raised']:
+                raise err
```
Keep this as `raise`; it preserves the stack trace.

jolyonb commented on mitxgraders/formulagrader/matrixgrader.py:
```diff
@@ -59,3 +85,78 @@ def check_response(self, answer, student_input, **kwargs):
+
+        Assumes comparer_params is just the single expected answer wrapped in a list.
+        """
+        expected_input =comparer_params[0]
```
Linting nit: space after `=`!

jolyonb commented on mitxgraders/formulagrader/matrixgrader.py:
```diff
@@ -59,3 +85,78 @@ def check_response(self, answer, student_input, **kwargs):
+            # in numpy, scalars have empty tuples as their shapes
+            shape = tuple() if isinstance(expected_input, Number) else expected_input.shape
+            utils.validate_shape(student_input, shape)
+        except ShapeError as err:
```
Can we move this try/except block into MatrixGrader, so that comparers can just raise a ShapeError and it's treated appropriately? Open to suggestions here.

jolyonb commented:
Looking much better. Few very minor comments, one small refactor comment.

ChristopherChudzicki commented on mitxgraders/formulagrader/matrixgrader.py:
```diff
@@ -58,4 +90,78 @@ def check_response(self, answer, student_input, **kwargs):
                 raise
             else:
                 return {'ok': False, 'msg': err.message, 'grade_decimal': 0}
+        except InputTypeError as err:
```
Re old comment:
> Can we move this try/except block into MatrixGrader, so that comparers can just raise a ShapeError and it's treated appropriately? Open to suggestions here.

I moved the try/except block out of the comparer, but in order to do what we want (if I understood correctly) we need a new error type. Reason is that ShapeErrors happen for two reasons:

- Student input contains shape errors like `"[[1, 2, 3]] - 2`
- Author encounters a shape error when subtracting `comparer_params[0] - student_input`. In this case it would be confusing to student to see "Cannot add/subtract matrix and scalar", since student doesn't "see" any subtraction happening. So these errors are re-raised as InputTypeErrors.

jolyonb commented:
LGTM 👍

jolyonb commented:
I think these last few changes have made implementing #74 much easier now, too.
